File: 1cuatri/ED/final/2019-feb/haskell/DataStructures/Graph/DictionaryWeightedGraph.hs
----------------------------------------------
-- Estructuras de Datos.  2018/19
-- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
-- Escuela Técnica Superior de Ingeniería en Informática. UMA
--
-- Examen 4 de febrero de 2019
--
-- ALUMNO/NAME:
-- GRADO/STUDIES:
-- NÚM. MÁQUINA/MACHINE NUMBER:
--
-- Weighted Graph implemented by using a dictionary from
-- sources to another dictionary from destinations to weights
----------------------------------------------

module DataStructures.Graph.DictionaryWeightedGraph
  ( WeightedGraph
  , WeightedEdge(WE)
  , empty
  , isEmpty
  , mkWeightedGraphEdges
  , addVertex
  , addEdge
  , vertices
  , numVertices
  , edges
  , numEdges
  , successors
  ) where

import Data.List(nub, intercalate)

import qualified DataStructures.Dictionary.AVLDictionary as D

data WeightedEdge a w  = WE a w a deriving Show

instance (Eq a, Eq w) => Eq (WeightedEdge a w) where
  WE u w v == WE u' w' v' = (u==u' && v==v' || u==v' && v==u')
                              && w == w'

instance (Eq a, Ord w) => Ord (WeightedEdge a w) where
  compare (WE _ w _) (WE _ w' _) = compare w w'

data WeightedGraph a w  = WG (D.Dictionary a (D.Dictionary a w))

empty :: WeightedGraph a w
empty = WG D.empty

addVertex :: (Ord a) => WeightedGraph a w -> a -> WeightedGraph a w
addVertex g@(WG d1) v
  | D.isDefinedAt v d1  = g
  | otherwise           = (WG (D.insert v (D.empty) d1))

addEdge :: (Ord a, Show a) => WeightedGraph a w -> a -> a -> w -> WeightedGraph a w
addEdge g@(WG d1) src dst w 
  | not ((D.isDefinedAt src d1) && (D.isDefinedAt dst d1))  = error "DictionaryWeightedGraph addEdge: vertex not in graph"
  | otherwise                                               = WG d_sol
    where
      d_src     = parseMaybe (D.valueOf src d1)
      new_d_src = D.insert dst w d_src
      d_sol     = D.insert src new_d_src d1

parseMaybe :: Maybe a -> a
parseMaybe (Just x) = x
parseMaybe Nothing = error "parseMaybe: tried to parse a Nothing"

edges :: (Ord a, Eq a, Eq w) => WeightedGraph a w -> [WeightedEdge a w]
edges g@(WG d1) = aux xs []
  where
    xs = D.keys d1  -- vertex

    -- Selecciona los vertices de los cuales hay que sacar aristas
    aux [] sol     = sol
    aux (v:vs) sol = aux vs sol_v
      where
        dic = parseMaybe (D.valueOf v d1)
        sol_v = aux2 v (ws) sol -- Añade aristas con src = v a la solución
        (ws) = D.keysValues dic

        -- Añade aristas con src = v a la solución
        aux2 v [] sol = sol
        aux2 v ((dst,w):ws) sol = (aux2 v ws sol) ++ [WE v w dst]

    

successors :: (Ord a, Show a) => WeightedGraph a w -> a -> [(a,w)]
successors g@(WG d1) v
  | not (D.isDefinedAt v d1)  = error "DictionaryWeightedGraph successors: vertex not in graph"
  | otherwise                 = aux suc dic
    where
      dic = parseMaybe (D.valueOf v d1)
      suc = D.keys dic

      aux [] dic = []
      aux (s:suc) dic = aux suc dic ++ [(s,w)]
        where
          w = parseMaybe (D.valueOf s dic)


-- NO EDITAR A PARTIR DE AQUÍ    
-- DON'T EDIT ANYTHING BELOW THIS COMMENT

vertices :: WeightedGraph a w -> [a]
vertices (WG d) = D.keys d

isEmpty :: WeightedGraph a w -> Bool
isEmpty (WG d) = D.isEmpty d

mkWeightedGraphEdges :: (Ord a, Show a) => [a] -> [WeightedEdge a w] -> WeightedGraph a w
mkWeightedGraphEdges vs es = wg'
  where
    wg = foldl addVertex empty vs
    wg' = foldr (\(WE u w v) wg -> addEdge wg u v w) wg es

numVertices :: WeightedGraph a w -> Int
numVertices = length . vertices

-- he añadido Ord a aqui
numEdges :: (Ord a, Eq a, Eq w) => WeightedGraph a w -> Int
numEdges = length . edges

-- he añadido Ord a aqui
instance (Ord a, Eq a, Show a, Eq w, Show w) => Show (WeightedGraph a w) where
  show wg  = "DictionaryWeightedGraph("++vs++", "++as++")"
   where
    vs  = "("++ intercalate ", " (map show (vertices wg)) ++")"
    as  = "(" ++ intercalate ", " (map showEdge (edges wg)) ++ ")"
    showEdge (WE x w y)  = intercalate "-" [ show x, show w, show y ]

File: 1cuatri/ED/final/2019-feb/java/feb19/src/dataStructures/graph/DictionaryWeightedGraph.java
/**----------------------------------------------
 * -- Estructuras de Datos.  2018/19
 * -- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
 * -- Escuela Técnica Superior de Ingeniería en Informática. UMA
 * --
 * -- Examen 4 de febrero de 2019
 * --
 * -- ALUMNO/NAME:
 * -- GRADO/STUDIES:
 * -- NÚM. MÁQUINA/MACHINE NUMBER:
 * --
 * ----------------------------------------------
 */

package dataStructures.graph;

import java.util.Iterator;
import java.util.Objects;

import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;

import dataStructures.set.Set;
import dataStructures.set.HashSet;
import dataStructures.tuple.Tuple2;

public class DictionaryWeightedGraph<V, W extends Comparable<? super W>> implements WeightedGraph<V, W> {

    static class WE<V1, W1 extends Comparable<? super W1>> implements WeightedEdge<V1, W1> {

		V1 src, dst;
        W1 wght;

        WE(V1 s, V1 d, W1 w) {
            src = s;
            dst = d;
            wght = w;
        }

        public V1 source() {
            return src;
        }

        public V1 destination() {
            return dst;
        }

        public W1 weight() {
            return wght;
        }

        public String toString() {
            return "WE(" + src + "," + dst + "," + wght + ")";
        }

		public int hashCode() {
			
			return Objects.hash(src, dst, wght);
		}

		public boolean equals(Object obj) {
			
			if(obj instanceof DictionaryWeightedGraph.WE<?,?>){
                WE<V1,W1> edge = (WE<V1, W1>) obj;
                return src.equals(edge.source()) && dst.equals(edge.destination()) &&
                        src.equals(edge.destination()) && dst.equals(edge.source()) &&  // para considerar aristas invertidas
                        weight().equals(edge.weight());
            }
            return false;
		}
		
		public int compareTo(WeightedEdge<V1, W1> o) {
			return weight().compareTo(o.weight());
		}
    }

    /**
     * Each vertex is associated to a dictionary containing associations
     * from each successor to its weight
     */
    protected Dictionary<V, Dictionary<V, W>> graph;

    public DictionaryWeightedGraph() {
        graph = new HashDictionary<>();
    }


    public void addVertex(V v) {
        if(!graph.isDefinedAt(v)){
            graph.insert(v, new HashDictionary<>());
        }

    	// COMPLETAR

    }

    public void addEdge(V src, V dst, W w) {
        if(!(graph.isDefinedAt(src) && graph.isDefinedAt(dst))){
            throw new GraphException("DictionaryWeightedGraph addEdge: vertex not in graph");

        }
        graph.valueOf(src).insert(dst, w);
        graph.valueOf(dst).insert(src,w);

    	// COMPLETAR

    }

    public Set<Tuple2<V, W>> successors(V v) {
        if(!graph.isDefinedAt(v)){
            throw new GraphException("DictionaryWeightedGraph successors: vertex not in graph");
        }
        Set<Tuple2<V,W>> s = new HashSet<>();
        Dictionary<V,W> dic = graph.valueOf(v);

//        for(V suc : dic.keys()){
//            s.insert(new Tuple2<>(suc, dic.valueOf(suc)));
//        }

        Iterator<V> it = dic.keys().iterator();
        while(it.hasNext()){
            V suc = it.next();
            s.insert(new Tuple2<>(suc, dic.valueOf(suc)));
        }

        return s;
    }


    public Set<WeightedEdge<V, W>> edges() {
        Set<WeightedEdge<V,W>> s = new HashSet<>();
//    	for(V vertex : graph.keys()){
//            Dictionary<V,W> dic = graph.valueOf(vertex);
//            for(V vertexConnected : dic.keys()){
//                W weight = dic.valueOf(vertexConnected);
//                s.insert(new WE<>(vertex, vertexConnected, weight));
//            }
//        }

        Iterator<V> it = graph.keys().iterator();
        while(it.hasNext()){
            V vertex = it.next();
            Dictionary<V,W> dic = graph.valueOf(vertex);

            Iterator<V> it2 = dic.keys().iterator();
            while(it2.hasNext()){
                V vertexConnected = it2.next();
                W weight = dic.valueOf(vertexConnected);
                s.insert(new WE<>(vertex, vertexConnected, weight));
            }
        }
    	
        return s;
    }






    /** DON'T EDIT ANYTHING BELOW THIS COMMENT **/


    public Set<V> vertices() {
        Set<V> vs = new HashSet<>();
        for (V v : graph.keys())
            vs.insert(v);
        return vs;
    }


    public boolean isEmpty() {
        return graph.isEmpty();
    }

    public int numVertices() {
        return graph.size();
    }


    public int numEdges() {
        int num = 0;
        for (Dictionary<V, W> d : graph.values())
            num += d.size();
        return num / 2;
//        return num;
    }


    public String toString() {
        String className = getClass().getSimpleName();
        String s = className + "(vertices=(";

        Iterator<V> it1 = vertices().iterator();
        while (it1.hasNext())
            s += it1.next() + (it1.hasNext() ? ", " : "");
        s += ")";

        s += ", edges=(";
        Iterator<WeightedEdge<V, W>> it2 = edges().iterator();
        while (it2.hasNext())
            s += it2.next() + (it2.hasNext() ? ", " : "");
        s += "))";

        return s;
    }
}